// smartux-client/build/config/main.webpack.js
let webpack = require('webpack');
let ionicWebpackFactory = require(process.env.IONIC_WEBPACK_FACTORY);

module.exports = {
  entry: process.env.IONIC_APP_ENTRY_POINT,

  output: {
    path: '{{BUILD}}',
    publicPath: 'build/',
    filename: 'main.js',
    devtoolModuleFilenameTemplate: ionicWebpackFactory.getSourceMapperFunction()
  },

  devtool: process.env.IONIC_SOURCE_MAP_TYPE,

  resolve: {
    symlinks: false,
    extensions: ['.ts', '.js', '.json'],
    modules: [process.env.IONIC_ROOT_DIR + '/node_modules'],
    alias: {
      app: process.env.IONIC_ROOT_DIR + '/src/app',
      mocks: process.env.IONIC_ROOT_DIR + '/src/mocks'
    }
  },

  module: {
    loaders: [
      {
        test: /\.json$/,
        loader: 'json-loader'
      },
      {
        test: /\.ts$/,
        loader: process.env.IONIC_WEBPACK_LOADER
      },
      {
        test: /\.js$/,
        loader: process.env.IONIC_WEBPACK_TRANSPILE_LOADER
      }
    ]
  },

  plugins: [
    new webpack.DllReferencePlugin({
      context: '.',
      manifest: require(process.env.IONIC_ROOT_DIR + '/www/build/vendors-manifest.json')
    }),
    new webpack.DllReferencePlugin({
      context: '.',
      manifest: require(process.env.IONIC_ROOT_DIR + '/www/build/components-manifest.json')
    })
  ],

  node: {
    fs: 'empty',
    net: 'empty',
    tls: 'empty'
  }
};
