// smartux-client/src/services/navigation.service.ts
import { Injectable } from '@angular/core';
import { App } from 'ionic-angular';
import { config } from '../config';
import { AppConfig, AppScreens } from '../types';
import { AppConfigService } from './app-config.service';
import { DeviceService } from './device.service';
import { ScreenService } from './screen.service';
import { UiService } from './ui.service';

@Injectable()
export class NavigationService {
  appConfig: AppConfig;
  appScreens: AppScreens;

  constructor(
    private screenService: ScreenService,
    private app: App,
    private deviceService: DeviceService,
    private uiService: UiService,
    private appConfigService: AppConfigService
  ) {}

  async initialize(screens: AppScreens): Promise<void> {
    this.appScreens = screens;
    this.appConfig = await this.appConfigService.get();
  }

  async go(screenId: string, animate: boolean = true): Promise<any> {
    if (!this.appScreens[screenId]) {
      let message = config.messages.missingScreen.replace('{{name}}', screenId);
      return this.uiService.alert('Missing screen', message, []);
    }

    let viewType = this.deviceService.getViewType();
    let component = this.appScreens[screenId][viewType];
    if (!component) {
      component = this.appScreens[screenId][this.appConfig.defaultLayout];
    }

    let currentComponent = this.screenService.getCurrentComponentName();
    if (currentComponent === component.name) {
      return; // do not make any transitions if this is the same component
    }

    let screen = await this.screenService.getById(screenId);
    if (screen.modal) {
      return this.uiService.modal(component);
    } else {
      return this.app.getActiveNavs()[0].setRoot(component, {}, {animate: animate, direction: 'forward'});
    }
  }

  async openStartScreen(): Promise<any> {
    let screenId = await this.screenService.getStartScreenId();
    return this.go(screenId);
  }

  openUrl(url: string) {
    window.open(url, '_system', 'location=yes,clearcache=yes,toolbar=no');
  };

}
