// smartux-client/src/bootstrap.ts
import { InjectionToken } from '@angular/core';
import { ApiService } from './api/api.service.abstract';
import { PmClient } from './pm-rpc/client';
import { AppService } from './services/app.service';
import { NavigationService } from './services/navigation.service';
import { AppHooks, AppMocks, AppScreens, TBootstrap } from './types';
import { MockService } from './services/mock.service';

export const Bootstrap = new InjectionToken<TBootstrap>('Bootstrap');

export let bootstrapProvider = {
  provide: Bootstrap,
  deps: [AppService, NavigationService, ApiService, MockService, PmClient],
  useFactory: (
    appService: AppService,
    navigationService: NavigationService,
    apiService: ApiService,
    mockService: MockService,
    pmClient: PmClient
  ) => {
    return async (hooks: AppHooks, screens: AppScreens, mocks: AppMocks) => {

      appService.initialize(hooks);
      await navigationService.initialize(screens);

      mockService.initialize(mocks);

      pmClient.initialize(window.parent);
      pmClient.createNotification('ready').call(null);

      let params = await hooks.getServerInitParams();
      return apiService.initialize(params);
    };
  }
};
