// smartux-client/src/config.ts
export let config = {
  ionic: {
    menuType: 'overlay',
    swipeBackEnabled: false
  },

  rpc: {
    defaultHost: 'localhost',
    defaultPort: 3000,
    loadingHidden: []
  },

  refreshTokenInterval: 2 * 60 * 1000,

  defaultEnv: 'production',

  loadingTimeout: 30000, // How long before 'Log an Error' shows up

  keepAliveTimeout: 0, // How often to send heartbeat for HTTP servers ("0" means disabled).

  ignoreExceptions: false,  // Don't show Javascript exceptions to the user

  noLogExceptions: false,   // Don't log Javascript exceptions to the server

  messages: {
    longLoading: 'The operation is taking a long time. Continue waiting?',
    connectionLost: 'Connection to the server has been lost.',
    connecting: 'Establishing session...',
    init: 'Initializing...',
    alreadyConnected: 'You have been connected from another device',
    doReconnect: 'Do you want to reconnect?',
    jsErrorTitle: 'Javascript Exception',
    responseError: 'An error occurred when connecting to the server.',
    missingScreen: 'The requested screen "{{name}}" does not exist. Try to rebuild the app or fix the navigation.'
  }
};
