// smartux-client/src/types.ts
import { Page } from 'ionic-angular/navigation/nav-util';

export interface TBootstrap {
  (hooks: AppHooks, screens: AppScreens, mocks: AppMocks): Promise<any>;
}

export interface Dictionary<T> {
  [key: string]: T;
}

export interface AppHooks {
  getServerInitParams(): Promise<any>

  initializeUI(params: any): void

  overrideStateHandler(oldScreen: string, newScreen: string, data: any): string
}

export interface AppScreens {
  [key: string]: Dictionary<Page>
}

export interface AppMocks {
  [key: string]: any
}

export interface ScreenData {
  id: string
  name: string
  modal?: boolean
  screens?: ScreenData[]
  mockSet?: string
  startScreen?: string
}

export interface ScreenModel {
  type: string
  comment: string
  properties?: ModelPropsMap
  items?: ModelProp
  params?: ModelPropsMap
}

export interface ModelProp {
  type: string
  properties?: ModelPropsMap
  items?: ModelProp
  update?: {
    type: string
  }
}

export interface ModelPropsMap {
  [name: string]: ModelProp
}

export interface ScreenModelsMap {
  [key: string]: ScreenModel
}

export interface Params {
  [key: string]: any
}

export interface SelectData {
  selected: string | string[]
  options: { label: string, value: string }[]
}

export interface DiffObjectPath {
  field: string,
  isArray: boolean,
  index?: number,
  remainingPath: string
}

export interface AppConfig {
  name: string,
  defaultLayout: string,
  previewToken: string,
  isAppTemplate: boolean,
  id: string,
  clientVersion: string,
  hidden: boolean,
  isBuiltInTemplate: boolean
}

export interface PowwowLoginConfig {
  cbUrl: string,
  defaultCbUrl: string,
  appVersion: string
}
