// smartux-client/src/api/powwow-login-api.service.ts
import { Injectable, Injector } from '@angular/core';
import { BehaviorSubject } from 'rxjs/Rx';
import { config } from '../config';
import { BaseScreen } from '../screen';
import { AppService } from '../services/app.service';
import { DataService } from '../services/data.service';
import { PowwowLoginService } from '../services/powwow-login.service';
import { UiService } from '../services/ui.service';
import { WsService } from '../services/ws.service';
import { ApiService } from './api.service.abstract';
import { HandlerService } from './handler.service';

@Injectable()
export class PowwowLoginApiService extends ApiService {
  private data: any;

  data$: BehaviorSubject<any> = new BehaviorSubject({});

  constructor(
    private wsService: WsService,
    private handlerService: HandlerService,
    private uiService: UiService,
    private dataService: DataService,
    private injector: Injector
  ) {
    super();
  }

  initialize(options: any): Promise<void> {
    this.uiService.showLoading();
    let appService = this.injector.get(AppService);
    let powwowLoginService = this.injector.get(PowwowLoginService);
    powwowLoginService.initialize().then(() => {
      appService.getAppSettings(false).then(() => {
        appService.establishConnection();
      });
    });

    this.wsService.onMessage(async event => {
      let rpcRequest = JSON.parse(event.data);
      if (!rpcRequest.method) {
        return;
      }
      let screenId = await this.handlerService.handleResponse(rpcRequest.method, rpcRequest.params);
      if (rpcRequest.method == 'SetState.Login' || rpcRequest.method == 'App.Login') {
        appService.hideLogin = true;
        appService.establishConnection();
      } else {
        // before we load next screen after login - check if our app sources are updated
        // (only for mobile devices)
        if ((rpcRequest.method.indexOf('SetState') > -1 || rpcRequest.method.indexOf('App.') > -1)
          && this.injector.get(PowwowLoginService).checkIfAppRequiresRefresh()) {
          window.location.hash = 'updateSources';
          window.location.reload(true);
        } else {
          this.data = await this.dataService.processResponse(this.data, rpcRequest.params, screenId);
          this.data$.next(this.data);
        }
      }
    });
    return Promise.resolve();
  }

  request(method: string, params: any, context?: BaseScreen): Promise<void> {
    if (this.wsService.isConnected()) {
      // send JSON-RPC message if WS connection is open:
      this.wsService.send({
        jsonrpc: '2.0',
        method: method,
        params: params
      });
    } else {
      // show error message if WS connection is closed:
      let alert = this.uiService.alert(null, config.messages.connectionLost);
      alert.onDidDismiss(() => window.history.go(0));
    }
    return Promise.resolve();
  }

}
